File: Data Structures & Algorithms/merge-k-sorted-linked-lists/submission-1.rs

impl Solution {
    pub fn merge_k_lists(lists: Vec<Option<Box<ListNode>>>) -> Option<Box<ListNode>>
    {
        fn take_next<'a, I>(
            mut lists: I,
        ) -> Option<Box<ListNode>>
        where
            I: Iterator<Item = &'a mut Option<Box<ListNode>>>,
        {
            let mut best_list: &mut Option<Box<ListNode>> = &mut None;

            while let Some(curr_list) = lists.next()
            {
                let Some(list_head) = curr_list
                else {
                    continue;
                };

                let Some(best_head) = best_list
                else {
                    best_list = curr_list;
                    continue;
                };

                if list_head.val < best_head.val {
                    best_list = curr_list;
                }
            }

            let Some(mut best_head) = best_list.take()
            else {
                return None;
            };

            *best_list = best_head.next.take();
            return Some(best_head);
        }

        let mut head = None;

        for list in lists
        {
            let mut lists = [head.take(),list];

            let Some(next) = take_next(lists.iter_mut())
            else {
                continue;
            };
            let mut curr = head.insert(next);

            while let Some(next) = take_next(lists.iter_mut()) {
                curr = curr.next.insert(next);
            }
        }
        return head;
    }
}

File: Data Structures & Algorithms/merge-two-sorted-linked-lists/submission-1.cpp

class Solution {
public:
    ListNode* mergeTwoLists(ListNode* list1, ListNode* list2)
    {
        if (list1 == nullptr) {
            return list2;
        }
        if (list2 == nullptr) {
            return list1;
        }

        auto const take_next = [&]()
        {
            if (list1->val < list2->val) {
                auto const node = list1;
                list1 = list1->next;
                return node;
            } else {
                auto const node = list2;
                list2 = list2->next;
                return node;
            }
        };

        auto const head = take_next();
        auto curr = head;

        while (list1 != nullptr && list2 != nullptr) {
            curr->next = take_next();
            curr = curr->next;
        }

        if (list1 == nullptr) {
            curr->next = list2;
        }
        if (list2 == nullptr) {
            curr->next = list1;
        }

        return head;
    }
};
